src/server.c

```diff
@@ -840,15 +840,15 @@ int clientsCronHandleTimeout(client *c, mstime_t now_ms) {
  *
  * The function always returns 0 as it never terminates the client. */
 int clientsCronResizeQueryBuffer(client *c) {
-    size_t querybuf_size = sdsAllocSize(c->querybuf);
+    size_t querybuf_size = sdsalloc(c->querybuf);
     time_t idletime = server.unixtime - c->lastinteraction;
 
     /* There are two conditions to resize the query buffer:
      * 1) Query buffer is > BIG_ARG and too big for latest peak.
      * 2) Query buffer is > BIG_ARG and client is idle. */
     if (querybuf_size > PROTO_MBULK_BIG_ARG &&
-         ((querybuf_size/(c->querybuf_peak+1)) > 2 ||
-          idletime > 2))
+        ((c->querybuf_peak && (querybuf_size/(c->querybuf_peak+1)) > 2) ||
+         idletime > 2))
```

Comment:
i'm not certain this is right.
on one hand, if we keep the code as it is now, if the client stops being active, it'll indeed resize the buffer on the next call to `clientsCronResizeQueryBuffer`, which, in case there are only 5 clients connected (see CLIENTS_CRON_MIN_ITERATIONS), will be on the next cron (by default after 100ms).

on the other hand, even with this change the peak mechanism is ineffective.
i think we need some cyclic buffer or alike to remember older peaks and let them retire after a fixed amount of time, rather than depend on server cron rate.

@yoav-steinberg WDYT?

Reply:
IIUC a few things come to mind:
1) The cron will shrink the qb in case the peak in the last time cycle was less than the allocated size. This means that zeroing the peak just doesn't make sense. At the moment of zeroing we already have the **current** qb size, which is more accurate than setting it to 0 and hoping we won't be idle. If we were idle then the comparison to 0 doesn't make sense. Why not reset it to the current qb size? 
2) I think holding a window of multiple peaks is the same as simply making sure the peak is reset every N cron cycles instead of every cron cycle. We just need the maximum in the last T time and not a window of multiple maximums. We can make this change but what T should we use? I suggest having T equal to the cron rate, in other words make no change.

Reply:
There's a difference between the peak of the last T time, and resetting it every N crons.
the difference is that right after you've reset it, you know nothing on the recent history.
so in my mind, resetting (to 0, or better yet to the current size) is only right after you taken an action to resize it.
if the reason we reset it is so that we have the recent peak and not the all-time peak, we need some cyclic buffer.

Reply:
@oranagra Right, you convinced me. But still I'm not sure we need to add this complexity. It might be enough just to reset the peak to the current qb used size every cron.

Reply:
i realize now (also by looking at the comment in the code) that the condition for idle clients (2 seconds) is a completely separate condition than the peak memory condition.

so maybe we don't need a cyclic buffer to hold 2 seconds of peak history.
my concern with holding a single peak and resetting it once in a while, was that right after reset, we are completely blind (we don't have any peak), but since we don't actually read from this peak variable in the near future, maybe that's ok.

i.e. if we have many clients, we probe them once per second, each time we look at the peak since the last probing, and then reset the peak, so that on the next probing we'll know the peak of the last second.
if there are less clients, we may probe each on every cron tick (e.g. once in 100ms).

so i suppose that resetting it to the current usage (rather than 0), could be sufficient (and drop the new condition added in this PR)

so the end result first check these two conditions to.
1. resize only if the size > 32k
2. resize only if the free portion is more than 4k
3. resize only if one of the following is true:
    - client is idle for over 2 seconds
    - client has less than 50% compared to a peak of the recent 1 second (or less)

Reply:
@oranagra I agree with using the recent N second to determine whether to check.
But there is a problem, when idle 2 seconds, we actually need to detect 3 seconds of previous peak, if so,
we can only detect 3 seconds of peaks to be valid.

Reply:
@sundb i think we misunderstand each other.
in my last post i meant that the idle and peak are two completely independent checks.
the idle has 2 seconds timeout, and the peak has 1 second (or even less)